src/test/socket/connect/test_connect.rs

```diff
@@ -0,0 +1,526 @@
+/*
+ * The Shadow Simulator
+ * See LICENSE for licensing information
+ */
+
+use test_utils::AsPtr;
+
+struct ConnectArguments {
+    fd: libc::c_int,
+    addr: Option<libc::sockaddr_in>, // if None, a null pointer should be used
+    addr_len: libc::socklen_t,
+}
+
+fn main() -> Result<(), String> {
+    // should we run only tests that shadow supports
+    let run_only_passing_tests = std::env::args().any(|x| x == "--shadow-passing");
+    // should we summarize the results rather than exit on a failed test
+    let summarize = std::env::args().any(|x| x == "--summarize");
+
+    let tests = if run_only_passing_tests {
+        get_passing_tests()
+    } else {
+        get_all_tests()
+    };
+
+    test_utils::run_tests(&tests, summarize)?;
+
+    println!("Success.");
+    Ok(())
+}
+
+fn get_passing_tests() -> std::collections::BTreeMap<String, test_utils::TestFn> {
+    #[rustfmt::skip]
+    let mut tests: Vec<(String, test_utils::TestFn)> = vec![
+        ("test_invalid_fd".to_string(),
+            Box::new(test_invalid_fd)),
+        ("test_non_existent_fd".to_string(),
+            Box::new(test_non_existent_fd)),
+        ("test_null_addr".to_string(),
+            Box::new(test_null_addr)),
+        ("test_short_len".to_string(),
+            Box::new(test_short_len)),
+        ("test_zero_len".to_string(),
+            Box::new(test_zero_len)),
+    ];
+
+    // tests to repeat for different socket options
+    for &sock_type in [libc::SOCK_STREAM, libc::SOCK_DGRAM].iter() {
+        for &flag in [0, libc::SOCK_NONBLOCK, libc::SOCK_CLOEXEC].iter() {
+            // add details to the test names to avoid duplicates
+            let append_args = |s| format!("{} <type={},flag={}>", s, sock_type, flag);
+
+            #[rustfmt::skip]
+            let more_tests: Vec<(String, test_utils::TestFn)> = vec![
+                (append_args("test_after_close"),
+                    Box::new(move || test_after_close(sock_type, flag))),
+                (append_args("test_interface_loopback"),
+                    Box::new(move || test_interface(sock_type, flag, libc::INADDR_LOOPBACK, None))),
+                (append_args("test_interface_loopback_any"),
+                    Box::new(move || test_interface(sock_type, flag, libc::INADDR_LOOPBACK, Some(libc::INADDR_ANY)))),
+                (append_args("test_interface_any"),
+                    Box::new(move || test_interface(sock_type, flag, libc::INADDR_ANY, None))),
+                (append_args("test_interface_any_any"),
+                    Box::new(move || test_interface(sock_type, flag, libc::INADDR_ANY, Some(libc::INADDR_ANY)))),
+                (append_args("test_double_connect_same_addr"),
+                    Box::new(move || test_double_connect(sock_type, flag, false))),
```

Comment:
The reader has to look pretty far away to understand what 'true' and 'false' means in this context. Consider annotating what bool parameters are at call sites (or avoiding boolean params in favor of an enum).

```
test_double_connect(sock_type, flag, /* change_address= */false)
```

or

```
enum ChangeAddress{ True, False };
...
test_double_connect(sock_type, flag, ChangeAddress::False)
```

Nice writeup about this issue here: https://abseil.io/tips/94

Reply:
Good idea, I went with the first since it's just a test, but for public interfaces the second approach seems like a good idea.